emcc.py

```diff
@@ -1157,14 +1157,28 @@ def get_last_setting_change(setting):
     # Set ASM_JS default here so that we can override it from the command line.
     shared.Settings.ASM_JS = 1 if shared.Settings.OPT_LEVEL > 0 else 2
 
+    # Remove the default _main function from shared.Settings.EXPORTED_FUNCTIONS.
+    # This happens before the user settings are applies.
+    if options.no_entry:
+      shared.Settings.EXPORTED_FUNCTIONS.remove('_main')
+
     # Apply -s settings in newargs here (after optimization levels, so they can override them)
     apply_settings(settings_changes)
 
-    if options.no_entry and '_main' in shared.Settings.EXPORTED_FUNCTIONS:
-      shared.Settings.EXPORTED_FUNCTIONS.remove('_main')
+    if options.no_entry or '_main' not in shared.Settings.EXPORTED_FUNCTIONS:
+      shared.Settings.HAS_MAIN = 0
```

Comment:
it looks like this line is where the value is set, and it's 1 otherwise. The meaning is not exactly "has" main maybe, but more "main exists and is exported"? But I'm not sure I have a better name idea. But maybe we could clarify this in the settings file where it's documented?

What happens if `--no-entry` but main exists? Does lld error?

Reply:
Yes, it does mean "main exists and is exported", but that seems to be the logical equivalent "program has a main function" and therefore calling it HAS_MAIN makes sense to me.

If have a main but you you don't export it we have special check for that which happens even before we link: see https://github.com/emscripten-core/emscripten/blob/99d6f4f34625dde97902684981e8d081e3904734/tools/system_libs.py#L1461

Reply:
I renamed the settings to EXPECT_MAIN and reduced the size of this change a fair amount.  Should be simpler to review now.